go-controller/pkg/ovn/pod_selector_address_set_test.go

```diff
@@ -0,0 +1,541 @@
+		refPodSelAS, _ := addressset.GetDbObjsForAS(refPodSelIDs, []net.IP{net.ParseIP("1.1.1.4")})
+		refPodSelAS.UUID = refPodSelAS.Name + "-UUID"
+		podSelACL := libovsdbops.BuildACL(
+			"podSelACL",
+			nbdb.ACLDirectionFromLport,
+			types.EgressFirewallStartPriority,
+			fmt.Sprintf("ip4.src == {$%s} && outport == @a13757631697825269621", refPodSelAS.Name),
+			nbdb.ACLActionAllowRelated,
+			types.OvnACLLoggingMeter,
+			"",
+			false,
+			nil,
+			map[string]string{
+				"apply-after-lb": "true",
+			},
+		)
+		podSelACL.UUID = "podSelACL-UUID"
+
+		initialDb := []libovsdbtest.TestData{
+			staleNetpolAS,
+			unusedPodSelAS,
+			refNetpolAS,
+			netpolACL,
+			refPodSelAS,
+			podSelACL,
+		}
+		dbSetup := libovsdbtest.TestSetup{NBData: initialDb}
+		fakeOvn.startWithDBSetup(dbSetup)
+
+		err := fakeOvn.controller.cleanupPodSelectorAddressSets()
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		finalDB := []libovsdbtest.TestData{
+			refNetpolAS,
+			netpolACL,
+			refPodSelAS,
+			podSelACL,
+		}
+		gomega.Eventually(fakeOvn.nbClient).Should(libovsdbtest.HaveData(finalDB))
+	})
+	ginkgo.It("reconciles a completed and deleted pod whose IP has been assigned to a running pod", func() {
+		namespace1 := *newNamespace(namespaceName1)
+		nodeName := "node1"
+		podIP := "10.128.1.3"
+		peer := knet.NetworkPolicyPeer{
+			PodSelector: &metav1.LabelSelector{},
+		}
+
+		fakeOvn.startWithDBSetup(initialDB,
+			&v1.NamespaceList{
+				Items: []v1.Namespace{
+					namespace1,
+				},
+			},
+		)
+
+		_, _, _, err := fakeOvn.controller.GetPodSelectorAddressSet(
+			peer.PodSelector, peer.NamespaceSelector, namespace1.Name, "backRef")
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		// Start a pod
+		completedPod, err := fakeOvn.fakeClient.KubeClient.CoreV1().Pods(namespace1.Name).
+			Create(
+				context.TODO(),
+				newPod(namespace1.Name, "completed-pod", nodeName, podIP),
+				metav1.CreateOptions{})
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		// pod should be added to the address set
+		eventuallyExpectAddressSetsWithIP(fakeOvn, peer, namespace1.Name, podIP)
+
+		// Spawn a pod with an IP address that collides with a completed pod (we don't watch pods in this test,
+		// therefore the same ip is allowed)
+		_, err = fakeOvn.fakeClient.KubeClient.CoreV1().Pods(namespace1.Name).
+			Create(
+				context.TODO(),
+				newPod(namespace1.Name, "running-pod", nodeName, podIP),
+				metav1.CreateOptions{})
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		// Mark the pod as Completed, so delete event will be generated
+		completedPod.Status.Phase = v1.PodSucceeded
+		completedPod, err = fakeOvn.fakeClient.KubeClient.CoreV1().Pods(completedPod.Namespace).
+			Update(context.TODO(), completedPod, metav1.UpdateOptions{})
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		// make sure the delete event is handled and address set is not changed
+		time.Sleep(100 * time.Millisecond)
```

Comment:
would be nice if there was a better way than a sleep

Reply:
would be nice, but I don't think it is possible, because retry framework for pod selector address set handlers is not stored anywhere (check `addNamespacedPodHandler`), therefore we can't check if the event was already handled or not. And this event should have no side effects, so I don't see any other way :(